src/Microsoft.AspNet.OData/Query/PlainTextODataQueryOptionsParser.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System.IO;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.OData;
+using Microsoft.AspNet.OData.Common;
+
+namespace Microsoft.AspNet.OData.Query
+{
+    [SuppressMessage("Microsoft.Naming", "CA1702:CompoundWordsShouldBeCasedCorrectly", Justification="Spelling correct in this context")]
+    public partial class PlainTextODataQueryOptionsParser
+    {
+        /// <inheritdoc/>
+        public string Parse(Stream requestStream)
+        {
+            string queryString = string.Empty;
+
+            MemoryStream memoryStream = new MemoryStream();
+            TextReader reader;
+
+            try
+            {
+                // Reset request stream position if possible
+                if (requestStream.CanSeek)
+                {
+                    requestStream.Position = 0;
+                }
+
+                requestStream.CopyTo(memoryStream);
+                memoryStream.Position = 0;
+                reader = new StreamReader(memoryStream);
+            }
```

Comment:
@habbes The reason for copying the content to a memory stream is because to read the content I'm using a reader and how they behave is that they dispose the underlying stream after they're done. I don't want that to happen here. The responsibility of disposing that request stream lies somewhere else.
The reason for attempting to rewind the stream if it supports seeking to make sure it's at position 0 before copying commences. If the stream position was (say) at position 13 and you called `CopyTo` (or `CopyAsync()`), the copying starts at that position.
It's also worth mentioning that not all streams support seeking.
Luckily, we intercept the request very early in the lifecycle such that we're almost always guaranteed that it's at position 0.

Reply:
I'm still wondering whether it's a good idea to rewind to position 0. We don't expect the request stream to have been read at this point. If someone did it, either they did it in error or they knew what they were doing. If they did it in error, I don't think we should attempt to fix that by rewinding it, I think it should just give incomplete results and the user would trace the issue back to their mistake, otherwise it will be harder to detect that mistake. If they did it intentionally for some obscure reason, then I don't think we should interfere with that decision. Maybe they have this hack where the payload has some random string at the beginning followed by the actual query options, so they decide to read that initial part of the stream in advance so that the odata parser can pick it up from where the query options actually start. I know that's very contrived and arbitrary scenario, but I think any scenario that intentionally leads the stream not to start at 0 would be. If it should be rewound, I think that should be the responsibility of the one who read it prematurely to rewind it.

Reply:
@habbes I don't mind removing it. The stream in `HttpRequest` doesn't support seeking either way. But just as an FYI, we actually do manipulate stream position quite significantly across OData repos:
- https://github.com/search?q=%22Stream.Position+%3D+%22+repo%3AOData%2FWebApi+path%3Asrc&type=Code&ref=advsearch&l=&l=
- https://github.com/search?q=%22Stream.Position+%3D+%22+repo%3AOData%2Fodata.net+path%3Asrc&type=Code&ref=advsearch&l=&l=
However, the execution flow in this case would support your argument

Reply:
@habbes Fixed